### PR TITLE
Encapsulate Render State

### PR DESCRIPTION
**Pull request by RobertBColton**
It should be very obvious what I am doing here. I am moving all of the namespace enigma:: scoped render state to a struct whose type is `RenderState` and whose members are public by default. This completes the first part of the work detailed in #1681 that is necessary to add dirty tracking. By having all of the render state encapsulated in a struct, we can do things like push and pop states as well as serialize the state to targets other than a GPU, such as including it in the `game_save` serialization.

I am not exactly sure of the performance implications of this change, but haven't noticed any degradation in any benchmarks. I have done extensive regression testing to ensure there's been no relapse in correctness either.

@JoshDreamland I think we also need to begin considering now what type of state belongs in this struct, and what type doesn't. For example, there is "render" state, such as circle precision, which is used by the general graphics functions, but does **NOT** need to be known by the GPU. However, it may still be useful to include in a `game_save` serialization and thus could still go into the new struct.

### CODE REVIEW

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSd3d.h [hunk 1]
```diff
@@ -19,88 +19,119 @@
...
 
 struct Light {
   gs_scalar x=0, y=0, z=0, range=0;
   bool directional=false;
   int color=0;
 };
 
-extern int d3dLightsActive, d3dLightingAmbient;
-const Light& get_active_light(int id);
+struct RenderState {
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSd3d.h]: On that note... How big is this struct? We may want to rearrange it to keep it within a page.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Graphics_Systems/General/GSd3d.h]: RenderState struct Size in Visual Studio

Comment (JoshDreamland) [reply ENIGMAsystem/SHELL/Graphics_Systems/General/GSd3d.h]: Seems fine. Carry on. Note that if this gets out of hands, you can also reduce the size by changing the order of the elements in this struct. You're wasting a byte here and there by declaring three bools followed by a short, etc.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Graphics_Systems/General/GSd3d.h]: I never knew, thanks for that.

**Pull request closed without merging**